### PR TITLE
C#: Extract init only accessors from CIL

### PR DESCRIPTION
**Pull request by tamasvajk**
`init` accessors are normal `set` accessors in IL, except that there's a `modreq` applied to them. To extract this `modreq` or `modopt` this PR introduces a `ModifiedType`, which references the unmodified type, the modifier, and the isRequired flag. These types are not written to the trapfiles, which means that when a `ModifiedType` would be written to a trapfile, we need to get the underlying unmodified type and write that instead. This is a bit error prone, because I didn't yet find a clear specification on where these modifiers can show up, but it looks like there are not that many places (fields, properties, methods, parameters and function pointers). See first bullet point in [this remark](https://docs.microsoft.com/en-us/dotnet/api/system.runtime.compilerservices.isvolatile?view=net-5.0#remarks).

Still missing:
- [x] stats https://jenkins.internal.semmle.com/job/Stats/job/CSharp/23/
- [x] DB upgrade folder
- [x] differences job 
  - [x] old stats: https://jenkins.internal.semmle.com/job/Changes/job/CSharp-Differences/700/
  - [x] changed stats: https://jenkins.internal.semmle.com/job/Changes/job/CSharp-Differences/702/, https://jenkins.internal.semmle.com/job/Changes/job/CSharp-Differences/729/

### CODE REVIEW

File: csharp/ql/src/semmlecode.csharp.dbscheme [hunk 1]
```diff
@@ -1718,8 +1719,14 @@ cil_parameter_out(unique int id: @cil_parameter ref);
...
 
-cil_init_only_accessors(
-  unique int id: @cil_method ref);
+#keyset[id, modifier]
+cil_custom_modifiers(
+  int id: @cil_custom_modifier_receiver ref,
+  int modifier: @cil_type ref,
+  int kind: int ref); // modreq: 1, modopt: 0
+
+//ew Tuple("cil_custom_modifiers", receiver, modifier, isRequired ? 1 : 0);
```
Comment (tamasvajk) [inline csharp/ql/src/semmlecode.csharp.dbscheme:1728]: maybe remove this.

File: csharp/ql/src/semmle/code/cil/CustomModifierReceiver.qll [hunk 2]
```diff
@@ -0,0 +1,29 @@
+/**
+ * Provides a class to represent `modopt` and `modreq` declarations.
+ */
+
+private import CIL
+private import dotnet
+
+/**
+ * A class to represent entities that can recive custom modifiers. Custom modifiers can be attached to
```
Comment (tamasvajk) [inline csharp/ql/src/semmle/code/cil/CustomModifierReceiver.qll:9]: typo: `recive`

File: csharp/extractor/Semmle.Extraction.CIL/Entities/ModifiedType.cs [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+
+        public override Type? ContainingType => throw new NotImplementedException();
+
+        public override IEnumerable<Type> TypeParameters => throw new NotImplementedException();
+
+        public override int ThisTypeParameterCount => throw new NotImplementedException();
+
+        public override Type Construct(IEnumerable<Type> typeArguments) => throw new NotImplementedException();
+
+        public override string Name => Unmodified.Name + (IsRequired ? " modreq" : " modopt") + $"({Modifier.Name})";
```
Comment (hvitved) [inline csharp/extractor/Semmle.Extraction.CIL/Entities/ModifiedType.cs:36]:
```suggestion
        public override string Name => $"{Unmodified.Name} {IsRequired ? "modreq" : "modopt"}({Modifier.Name})";
```

File: csharp/extractor/Semmle.Extraction.CIL/Entities/ModifiedType.cs [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+
+        public override void WriteAssemblyPrefix(TextWriter trapFile) => Unmodified.WriteAssemblyPrefix(trapFile);
+
+        public override void WriteId(TextWriter trapFile, bool inContext)
```
Comment (hvitved) [inline csharp/extractor/Semmle.Extraction.CIL/Entities/ModifiedType.cs:40]: I reckon this should never be called? So `throw NotImplementedException` would be better.

File: csharp/extractor/Semmle.Extraction.CIL/Entities/ModifiedType.cs [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+
+        public override IEnumerable<Type> TypeParameters => throw new NotImplementedException();
+
+        public override int ThisTypeParameterCount => throw new NotImplementedException();
+
+        public override Type Construct(IEnumerable<Type> typeArguments) => throw new NotImplementedException();
+
+        public override string Name => Unmodified.Name + (IsRequired ? " modreq" : " modopt") + $"({Modifier.Name})";
+
+        public override void WriteAssemblyPrefix(TextWriter trapFile) => Unmodified.WriteAssemblyPrefix(trapFile);
```
Comment (hvitved) [inline csharp/extractor/Semmle.Extraction.CIL/Entities/ModifiedType.cs:38]: Should this ever be called?

File: csharp/extractor/Semmle.Extraction.CIL/Entities/SignatureDecoder.cs [hunk 6]
```diff
@@ -138,21 +139,28 @@ public void WriteId(TextWriter trapFile, GenericContext gc)
...
+                this.modifier = modifier;
+                this.isRequired = isRequired;
             }
 
             public void WriteId(TextWriter trapFile, GenericContext gc)
             {
                 unmodifiedType.WriteId(trapFile, gc);
+                trapFile.Write(isRequired ? " modreq(" : " modopt(");
+                modifier.WriteId(trapFile, gc);
+                trapFile.Write(")");
```
Comment (hvitved) [inline csharp/extractor/Semmle.Extraction.CIL/Entities/SignatureDecoder.cs:157]: Why is this added?I thought modified types were not written to TRAP.

**Review by hvitved — CHANGES_REQUESTED** [review]
Looks great to me. Some questions.

**Comment by tamasvajk** [comment]
@hvitved I found some perf issue in the differences job. The below shows up in the roslyn snapshot1 log, but not in snapshot0:
```
ClassDoesNotImplementEquals.ql-11:System::getInvokedIEquatableEqualsMethod#bbf#shared ............. 20.5s
```

I also found the below line with a huge number of rows:
```
Created relation System::getInvokedIEquatableEqualsMethod#bbf#shared/3@a7bede with 62679616 rows.
```

Can we have a look together? I'don't know how to move forward with this issue.

File: csharp/extractor/Semmle.Extraction.CIL/Entities/ModifiedType.cs (hunk 4 above)
Comment (tamasvajk) [reply csharp/extractor/Semmle.Extraction.CIL/Entities/ModifiedType.cs:40]: True. It would be a bug, if it was called.

File: csharp/extractor/Semmle.Extraction.CIL/Entities/ModifiedType.cs (hunk 5 above)
Comment (tamasvajk) [reply csharp/extractor/Semmle.Extraction.CIL/Entities/ModifiedType.cs:38]: No, changed it to an exception throwing.

File: csharp/extractor/Semmle.Extraction.CIL/Entities/SignatureDecoder.cs (hunk 6 above)
Comment (hvitved) [reply csharp/extractor/Semmle.Extraction.CIL/Entities/SignatureDecoder.cs:157]: Thanks for the detailed explanation; I am convinced! :-) (I hadn't realized that the IDs written here were only part of method signatures.)

**Comment by hvitved** [comment]
Oh no, we are running into the BDD issue again 😱

**Comment by tamasvajk** [comment]
Latest diff job: https://jenkins.internal.semmle.com/job/Changes/job/CSharp-Differences/735/

**Comment by hvitved** [comment]
> Oh no, we are running into the BDD issue again 😱

Apparently fixed by my latest change 🤷‍♂️

**Review by hvitved — APPROVED** [review]
